src/check/constrain/generate/definition.rs

```diff
@@ -164,85 +165,85 @@ pub fn id_from_var(
     if let Some(expr) = expr {
         generate(expr, &env.is_expr(true), ctx, constr)?;
     }
-    let mut env = env.clone();
 
+    let mut env = env.clone();
     let identifier = Identifier::try_from(var.deref())?.as_mutable(mutable);
-    if let Some(ty) = ty {
-        for (f_name, (f_mut, name)) in match_name(&identifier, ty, var.pos)? {
-            let ty = Expected::new(var.pos, &Type { name: name.clone() });
-            constr.insert_var(&f_name);
-            env = env.insert_var(mutable && f_mut, &f_name, &ty, &constr.var_mapping);
-        }
-    } else {
-        let any = Expected::any(var.pos);
-        for (f_mut, f_name) in identifier.fields(var.pos)? {
-            constr.insert_var(&f_name);
-            env = env.insert_var(mutable && f_mut, &f_name, &any, &constr.var_mapping);
-        }
-    };
+    match (ty, expr) {
+        (Some(ty), Some(expr)) => {
```

Comment:
Have we tested type and expr?
Should the logic below for `Some(expr)` not also be here?
May be good reason for doing so but maybe just have a small comment then so we don't forget, because on the face of it this looks like a bug.

If it should be here, then better to have wildcard for `ty` and just have a small `if let Some(...) = ...`  binding for ty in the body.